33 referral referral code linking to solana wallet address

diamondly777 opened the pull request:
- Referral code generation UI integrated with backend.
- Connecting wallet with referral code
- UI confirmation is shown once the link is successful
- User can see and copy referral code/link.

All functions are integrated with the backend API endpoints and works well.

<img width="1907" height="995" alt="Screenshot_6" src="https://github.com/user-attachments/assets/31949165-4b04-4805-9e2d-46e6890e5ef9" />
<img width="914" height="1021" alt="Screenshot_4" src="https://github.com/user-attachments/assets/fe5c760d-a56f-4c01-b79e-132b5aa05506" />

redDwarf03 commented on lib/application/referral/provider.dart:
```diff
@@ -0,0 +1,141 @@
+  return ReferralRepositoryImpl(apiClient);
+});
+
+// Referral state notifier
```
You should separate notifier and state in `ui/views/xxx/bloc` folder to associate notifier management and UI View
Example : `lib/ui/views/forge_detail`

redDwarf03 commented on lib/ui/views/referral/layouts/components/referral_content.dart:
```diff
@@ -0,0 +1,649 @@
+import 'package:clones_desktop/application/referral/provider.dart';
```
Widget too long
please try to extract method to Widgets in `components` folder
See `lib/ui/views/forge_detail/layouts/components`

diamondly777 replied:
What have done:
I created separate component files in `lib/ui/views/referral/layouts/components/`:
  - `referral_header.dart`
  - `wallet_connection_section.dart`
  - `referral_code_card.dart`
  - `referral_stats_card.dart`
  - `referral_instructions_card.dart`

redDwarf03 commented on lib/ui/views/referral/layouts/components/referral_content.dart:
```diff
@@ -0,0 +1,649 @@
+                ),
+                const SizedBox(width: 12),
+                Expanded(
+                  child: Text(
```
Always use Assets.dart fonts like  Theme.of(context).textTheme.....

diamondly777 commented:
> <img alt="referal_program_icon" width="1024" height="1024" src="https://private-user-images.githubusercontent.com/68340133/472367649-02329551-2cdf-45df-b720-3dacff00057c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM4NjM1ODgsIm5iZiI6MTc1Mzg2MzI4OCwicGF0aCI6Ii82ODM0MDEzMy80NzIzNjc2NDktMDIzMjk1NTEtMmNkZi00NWRmLWI3MjAtM2RhY2ZmMDAwNTdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzMwVDA4MTQ0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0ZmFlZmNkOTA3YTYzN2JjNzdmZjliNDFjYjIyNDBmZDc1MjI5MDRjOGVmZjlmNzVkZmQ0NWI2ODE0MDBhMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ObEHL-97RqOZmJkRQBqs8GGAmmsQlOvah8O-hHp4UHw">

I updated the referral icon.

diamondly777 commented:
What have done:
- Removed macOS-only conditional compilation from commands module
- Added Windows and Linux permission functions alongside macOS
- Fixed PathBuf import to work on all platforms
- Added Windows-specific dependencies to Cargo.toml
- Added dev_tauri_desktop.ps1 for Windows desktop-only development
- Added dev_tauri_desktop.bat wrapper for easy execution
- Added dev_web_tauri.ps1 PowerShell version of web+tauri script
- Updated dev_web_tauri.bat to use PowerShell script
- Implemented temporary build directories to avoid file locks
- Added background process management with PowerShell jobs

redDwarf03 commented on lib/ui/components/wallet_button.dart:
```diff
@@ -34,18 +34,32 @@ class _WalletButtonState extends ConsumerState<WalletButton> {
+    
     try {
+      await ref.read(sessionNotifierProvider.notifier).getConnectionUrl();
+      if (!mounted) return;
```
Why condition on mounted here ?

diamondly777 replied:
<img width="1328" height="433" alt="image" src="https://github.com/user-attachments/assets/a5978469-939a-4476-b5e8-cdd62296e25a" />

redDwarf03 commented on lib/infrastructure/referral.repository.dart:
```diff
@@ -0,0 +1,77 @@
+import 'package:clones_desktop/domain/models/api/create_referral_request.dart';
```
Normally, you should have some warning due to `analysis_options.yaml` rules
For example in this class:
- Unused import
- Sort directive sections alphabetically
- 'ReferralRepositoryRef' is deprecated and shouldn't be used. Will be removed in 3.0. Use Ref instead.
- Unnecessary use of a 'double' literal.
- The operand can't be 'null', so the condition is always 'false'.
- The value of the argument is redundant 
- Missing a newline at the end of the file.

redDwarf03 commented on lib/ui/views/referral/bloc/provider.dart:
```diff
@@ -0,0 +1,116 @@
+    try {
+      debugPrint('🔍 [ReferralNotifier] Calling repository.createReferral...');
+      final repository = ref.read(referralRepositoryProvider);
+      final response = await repository.createReferral(walletAddress);
```
always `true` or throw an exception

redDwarf03 commented on lib/ui/views/referral/bloc/provider.dart:
```diff
@@ -0,0 +1,116 @@
+    try {
+      debugPrint('🔍 [ReferralNotifier] getReferralInfo called with walletAddress: $walletAddress');
+      final repository = ref.read(referralRepositoryProvider);
+      final response = await repository.getReferralInfo(walletAddress);
```
if the code is ok, you can remove `debugPrint`

diamondly777 replied:
removed

redDwarf03 commented on lib/ui/views/referral/bloc/provider.dart:
```diff
@@ -0,0 +1,116 @@
+        }
+        
+        // Construct the referral link using the referral code
+        final referralLink = 'https://clones-ai.com/ref/${response.data.referralCode}';
```
use `API_WEBSITE_URL` .env value because in dev environment, it couldn't work

redDwarf03 commented on lib/ui/views/referral/layouts/components/referral_code_card.dart:
```diff
@@ -0,0 +1,106 @@
+      width: double.infinity,
+      padding: const EdgeInsets.all(24),
+      decoration: BoxDecoration(
+        color: Colors.white.withOpacity(0.05),
```
Use `ClonesColor` class

redDwarf03 commented on lib/ui/views/referral/layouts/components/referral_content.dart:
```diff
@@ -0,0 +1,310 @@
+          const SizedBox(height: 32),
+          ElevatedButton(
+            onPressed: () {
+              debugPrint('🔍 [ReferralContent] Generate Referral Code button pressed');
```
if code ok, you czan remove `debugPrint`

redDwarf03 commented on lib/ui/views/referral/layouts/referral_view.dart:
```diff
@@ -0,0 +1,33 @@
+          width: 1,
+        ),
+      ),
+      child: ReferralContent(
```
not necessary to pass providers to `ReferralContent` widget.
you can use directly in the widget

redDwarf03 commented on lib/ui/views/referral/layouts/components/referral_content.dart:
```diff
@@ -0,0 +1,310 @@
+// Provider to track if we've already tried loading referral info
+final _hasTriedLoadingProvider = StateProvider<bool>((ref) => false);
+
+// Provider to track the previous wallet address
```
not necessary. Use `listen`method on `sessionNotifierProvider` and compare previous and next value

diamondly777 replied:
<img width="646" height="286" alt="image" src="https://github.com/user-attachments/assets/22032a7a-0596-4a2e-b831-522cbeaae06d" />

redDwarf03 commented on lib/ui/views/referral/layouts/components/referral_header.dart:
```diff
@@ -0,0 +1,27 @@
+        const SizedBox(width: 16),
+        Text(
+          'Referral Program',
+          style: Theme.of(context).textTheme.headlineMedium?.copyWith(
```
If `headlineMedium` is necessary here update `textTheme` in `main.dart`.
But this is the only time where we use `headlineMedium` in the app, so perhaps it's not the good theme

diamondly777 replied:
The code has been changed from headlineMedium to titleLarge, which is the correct approach since titleLarge is defined in the text theme and is more appropriate for this use case.

redDwarf03 commented on scripts/dev_web_tauri.sh:
```diff
@@ -1,15 +1,107 @@
-#!/bin/bash
+﻿#!/bin/bash
```
it doesn't work on macos now

```
(base) sse@macbookair desktop % ./scripts/dev_web_tauri.sh                         
./scripts/dev_web_tauri.sh: line 1: #!/bin/bash: No such file or directory
ðŸš€ Starting Flutter Web + Tauri Development
ðŸ–¥ï¸  Detected OS: macos
ðŸ§¹ Cleaning up existing processes...
🔍 Checking Flutter availability...
📱 Using Flutter command: flutter
🔍 Running Flutter doctor...
/Users/SSe/Library/Android/sdk/cmdline-tools/latest/bin/sdkmanager: line 173: test: : integer expression expected
[=======================================] 100% Computing updates...             
All SDK package licenses accepted.

📱 Starting Flutter Web development server...
📁 Flutter will output to: /Users/SSe/SSe/app/Viralmind/desktop/flutter.log
❌ Web server device not available. Available devices:
Found 3 connected devices:
  macOS (desktop)                 • macos                 • darwin-arm64   • macOS 15.5 24F74 darwin-arm64
  Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 15.5 24F74 darwin-arm64
  Chrome (web)                    • chrome                • web-javascript • Google Chrome 138.0.7204.169

Found 1 wirelessly connected device:
  iPhone SSe (wireless) (mobile) • 00008120-000A31CE0152201E • ios • iOS 18.5 22F76

Run "flutter emulators" to list and start any available device emulators.

If you expected another device to be detected, please run "flutter doctor" to diagnose potential issues. You may also try increasing the time to wait for connected devices with the "--device-timeout" flag. Visit https://flutter.dev/setup/ for troubleshooting tips.
ðŸ§¹ Cleaning up...
```

diamondly777 replied:
<img width="1443" height="438" alt="image" src="https://github.com/user-attachments/assets/a3dd71dc-86ab-43ac-997d-ed168824eb85" />

redDwarf03 commented on web/index.html:
```diff
@@ -1,48 +1,24 @@
-<!DOCTYPE html>
+﻿<!DOCTYPE html>
```
Why do you change flutter web index page ?

diamondly777 replied:
Restored original Flutter web index.html file

diamondly777 replied:
<img width="1377" height="668" alt="image" src="https://github.com/user-attachments/assets/38612b5b-c150-44ef-9c95-d15ffc2e2269" />

redDwarf03 replied:
it isn't restore as same as before.. but i will check this in other PR

diamondly777 commented on native/src-tauri/gen/schemas/acl-manifests.json:
it's the auto generated file by Tauri during the build process. It contains the Access Control List (ACL) manifest that defines all available permissions for Tauri plugins. I recommend to keep it at this moment.

diamondly777 commented on scripts/dev-windows.ps1:
```diff
@@ -0,0 +1,145 @@
+# Windows-specific development script for Tauri
```
its just a windows-specific script. it works well on windows

diamondly777 commented on lib/ui/components/wallet_button.dart:
```diff
@@ -34,18 +34,32 @@ class _WalletButtonState extends ConsumerState<WalletButton> {
+      if (!mounted) return;
+      
+      final session = ref.read(sessionNotifierProvider);
+      if (!mounted) return;
```
removed

diamondly777 commented on lib/ui/components/wallet_button.dart:
```diff
@@ -34,18 +34,32 @@ class _WalletButtonState extends ConsumerState<WalletButton> {
+      if (mounted) {
+        ScaffoldMessenger.of(context).showSnackBar(
+          SnackBar(
+            content: Text('Failed to connect wallet: $e'),
```
Added proper text style to SnackBar using Theme.of(context).textTheme.bodyMedium

diamondly777 commented on lib/ui/views/referral/bloc/provider.dart:
```diff
@@ -0,0 +1,116 @@
+      
+      if (response.success) {
+        // Check if the referral code is empty or null
+        if (response.data.referralCode == null || response.data.referralCode.isEmpty) {
```
Removed unnecessary null check for referralCode since it's a required field

diamondly777 commented on lib/ui/views/referral/layouts/components/referral_content.dart:
```diff
@@ -0,0 +1,310 @@
+class ReferralContent extends ConsumerWidget {
+  const ReferralContent({
+    super.key,
+    required this.sessionState,
```
Removed unnecessary parameters from ReferralContent widget

diamondly777 commented on scripts/dev_web_tauri.sh:
```diff
@@ -1,15 +1,107 @@
 set -euo pipefail
 
-echo "🚀 Starting Flutter Web + Tauri Development"
+echo "ðŸš€ Starting Flutter Web + Tauri Development"
```
Fixed emoji encoding issues in dev_web_tauri.sh

diamondly777 commented on lib/domain/models/api/create_referral_response.dart:
```diff
@@ -0,0 +1,17 @@
+@freezed
+class CreateReferralResponse with _$CreateReferralResponse {
+  const factory CreateReferralResponse({
+    required bool success,
```
removed success field from response models

diamondly777 commented on lib/ui/views/referral/bloc/provider.dart:
```diff
@@ -0,0 +1,116 @@
+          totalReferrals: response.data.totalReferrals,
+          totalRewards: response.data.totalRewards,
+          isActive: true, // Assume active if we get data
+          createdAt: DateTime.now(), // Backend doesn't provide this in stats
```
The code now uses DateTime.now() for createdAt since the backend doesn't provide it.
I will add createdAt to the backend response while working backend.

redDwarf03 replied:
please add a todo in the code
`// TODO(diamondly777): Backend doesn't provide this in stats`

diamondly777 commented on lib/ui/views/referral/bloc/provider.dart:
```diff
@@ -0,0 +1,116 @@
+    } catch (e) {
+      debugPrint('🔍 [ReferralNotifier] Exception in getReferralInfo: $e');
+      
+      // If it's a 404 error (wallet doesn't have referral code), show initial state instead of error
```
- Added import for ApiError class
- Replaced fragile string parsing (e.toString().contains('404')) with proper type checking (e is ApiError)
- Used specific status code checking (e.status == 404) instead of string matching

diamondly777 commented on lib/ui/views/referral/layouts/components/referral_instructions_card.dart:
```diff
@@ -0,0 +1,118 @@
+        ),
+      ),
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
```
<img width="1245" height="588" alt="image" src="https://github.com/user-attachments/assets/45ef9ad3-a572-4757-a5c4-5887f62edeb9" />

redDwarf03 replied:
sorry. my request was not clear.
when you want to add spacing between line in Column widget, you can declare `spacing` property
ex:
```
Column(
   spacing: 16,
    children: [
             .... 
  ],
```

diamondly777 replied:
<img width="592" height="236" alt="image" src="https://github.com/user-attachments/assets/cdc7aca5-abfd-4a20-802b-fac32b3389aa" />
- Added spacing: 16 to the Column widget properties
- Removed all ClonesSpacing.space16 widgets between the instruction steps

redDwarf03 commented on .vscode/settings.json:
```diff
@@ -1,3 +1,3 @@
 {
-  "dart.flutterSdkPath": "~/.asdf/installs/flutter/3.29.3-stable"
+  "dart.flutterSdkPath": "c:\\flutter"
```
We should add this file in `.gitignore`file because windows and macos don't use the same path

redDwarf03 commented on lib/domain/models/api/get_referral_info_response.dart:
```diff
@@ -0,0 +1,17 @@
+    required int totalReferrals,
+    required double totalRewards,
+    required String referralCode,
+    required List<dynamic> referrals,
```
instead of using dynamic properties, It could be better to create a freezed class based on backend schema 
```

const ReferralSchema = new mongoose.Schema<IReferral>(
  {
    referrerAddress: { 
      type: String, 
      required: true, 
      index: true 
    },
    referreeAddress: { 
      type: String, 
      required: true, 
      index: true,
      unique: true // Each wallet can only be referred once
    },
    referralCode: { 
      type: String, 
      required: true,
      index: true
    },
    referralLink: { 
      type: String, 
      required: true 
    },
    firstActionType: { 
      type: String, 
      required: true,
      enum: ['wallet_connect', 'first_task', 'first_submission', 'first_pool_creation']
    },
    firstActionData: { 
      type: mongoose.Schema.Types.Mixed 
    },
    onChainTxHash: { 
      type: String 
    },
    onChainSlot: { 
      type: Number 
    },
    rewardAmount: {
      type: Number,
      default: 0
    },
    rewardProcessed: {
      type: Boolean,
      default: false
    },
    status: { 
      type: String, 
      required: true, 
      enum: ['pending', 'confirmed', 'failed'],
      default: 'pending'
    }
  },
  {
    collection: 'referrals',
    timestamps: true
  }
);
```

diamondly777 replied:
- Updated GetReferralInfoResponse to use typed Referral instead of dynamic

diamondly777 commented on lib/ui/components/wallet_button.dart:
```diff
@@ -34,18 +34,30 @@ class _WalletButtonState extends ConsumerState<WalletButton> {
+          ),
+        );
+      }
       debugPrint('Failed to open external URL: $e');
```
Remove unnecessary debugPrint statement

diamondly777 commented on lib/ui/views/referral/layouts/components/referral_content.dart:
```diff
@@ -0,0 +1,277 @@
+            textAlign: TextAlign.center,
+          ),
+          const SizedBox(height: 32),
+          ElevatedButton(
```
Replace ElevatedButton with BtnPrimary

diamondly777 commented on lib/assets.dart:
```diff
@@ -34,11 +35,30 @@ class Assets {
   static const String menuForge = '${assetMenuFolder}forge_menu.png';
 }
 
+class ClonesSpacing {
```
The ClonesSpacing class is being used in referral_instructions_card.dart
